lib/search/packages.ex

```diff
@@ -0,0 +1,113 @@
+defmodule Search.Packages do
+  import Ecto.Query, warn: false
+  alias Search.Repo
+
+  alias Search.Packages.{Package, DocItem, DocFragment}
+  alias Search.{HexClient, ExDocParser}
+
+  @doc """
+  Adds the package to be indexed by the application.
+
+  If given a package name, adds the latest version of the package to the app. If given a `%HexClient.Release{}` adds
+  the specified release. Does not embed it yet.
+  """
+  def add_package(package_name) when is_binary(package_name) do
+    case HexClient.get_releases(package_name) do
+      {:ok, releases} ->
+        latest = HexClient.Release.latest(releases)
+        add_package(latest)
+
+      err ->
+        err
+    end
+  end
+
+  def add_package(%HexClient.Release{package_name: package_name, version: version} = release) do
+    version = Version.to_string(version)
+
+    with {:ok, docs} <- HexClient.get_docs_tarball(release),
+         {:ok, search_data} <- ExDocParser.extract_search_data(docs) do
+      Repo.transaction_with(fn ->
+        package =
+          case Repo.get_by(Package, name: package_name) do
+            nil ->
+              %Package{name: package_name, version: version}
+
+            existing ->
+              existing
+          end
+          |> Repo.preload(:doc_items)
+          |> Package.changeset(%{
+            version: version
+          })
+          |> Ecto.Changeset.put_assoc(:doc_items, [])
+
+        with {:ok, package} <- Repo.insert_or_update(package),
+             :ok <- create_items_from_package(package, search_data) do
+          {:ok, package}
+        else
+          {:error, _} = err ->
+            err
```

Comment:
this adds some strictness but I don't think it's worth it and so I'd just remove it.

```suggestion
```